### PR TITLE
CUDAAccelerator.setup_device: fix unrelated device init by matmul precision check

### LINKED ISSUE #21725: CUDAAccelerator.setup_device(device) initializes a different device (GPU 0)
### Bug description

# Problem

Despite passing values other than 0 as `devices` to `Trainer`, memory was allocated on device 0.

Passing a list of specific devices to `pytorch-lightning` ideally means that other devices are not used in any way.

# Investigation

`CUDAAccelerator.setup_device(device)` calls `_check_cuda_matmul_precision` before `torch.cuda.set_device(device)`.

`_check_cuda_matmul_precision` calls `torch.cuda._lazy_init` via
- `_check_cuda_matmul_precision`
- `_is_ampere_or_later`
- `torch.cuda.get_device_capability`
- `torch.cuda.get_device_properties`
- `torch.cuda._lazy_init`

# Is this a bug

It was not clear to me what the contract is regarding using devices not explicitly passed, but also not excluded via `CUDA_VISIBLE_DEVICES`.

A similar issue in `pytorch` convinced me this is a bug https://github.com/pytorch/pytorch/issues/149119

I also considered whether this was a pytorch-lightning bug or an upstream pytorch bug. I read `torch.cuda` code and it seems to me that  it is up to the caller to first call `torch.cuda.set_device` before other functions that require CUDA context.

# How to fix it

It seems to me that it's just a matter of reordering two lines, so that `torch.cuda.set_device(device)` is before `_check_cuda_matmul_precision`, which seems to be harmless: `_check_cuda_matmul_precision` is just for logging info for some devices.

I will submit a PR for this shortly.

### What version are you seeing the problem on?

v2.6, master

### Reproduced in studio

_No response_

### How to reproduce the bug

```python
Train with a trainer like


trainer = pytorch_lightning.Trainer(
    devices=[3,4,5,6],
    accelerator="gpu",
    strategy: "ddp_find_unused_parameters_true",
)


and check memory allocation via `nvidia-smi`
```

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
# - PyTorch Lightning Version 2.6.1
# - PyTorch Version 2.6.0
# - Python 3.10
# - Ubuntu
# - CUDA 12.4 /cuDNN 9
# 8 GPUs Nvidia A40
#  uv installation
```

</details>

### More info

_No response_

cc @ethanwharris @justusschock @lantiga

### PR DESCRIPTION
**Pull request by bm371613**
Without this fix, CUDAAccelerator.setup_device may initialize an unrelated device, via
- _check_cuda_matmul_precision
- _is_ampere_or_later
- torch.cuda.get_device_capability
- torch.cuda.get_device_properties
- torch.cuda._lazy_init

## What does this PR do?

Fixes #21725 

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Review by justusschock — COMMENTED** [review]
Great! Is there a way for us to test this?

**Comment by bm371613** [comment]
> Great! Is there a way for us to test this?

If you mean unit tests, I can see a simple test monkeypatching `torch.cuda` and checking that `set_device` is called before `_lazy_init`. If you don't mind digging into `torch.cuda` internals for testing.

**Comment by justusschock** [comment]
> If you mean unit tests, I can see a simple test monkeypatching torch.cuda and checking that set_device is called before _lazy_init. If you don't mind digging into torch.cuda internals for testing. 

As long it is local to the specific test and doesn't have side-effects to other tests, i think that sounds like a good approach

**Comment by bm371613** [comment]
> > If you mean unit tests, I can see a simple test monkeypatching torch.cuda and checking that set_device is called before _lazy_init. If you don't mind digging into torch.cuda internals for testing.
> 
> As long it is local to the specific test and doesn't have side-effects to other tests, i think that sounds like a good approach

I added the tests.

To ensure other tests are not affected I used `unittest.mock`. To ensure other tests do not affect the newly added test by initializing CUDA in the pytest process, I had to do process isolation. I found https://pypi.org/project/pytest-isolate/ but you do not have that as a dependency and  I suppose you may not want to add it. I marked the test with "skip" and an explanation and added another test to do the process spawning.

**Review by deependujha — APPROVED** [review]

**Comment by adhavan18** [comment]
@bhimrazy pointed out on #21767 that mine duplicates this one. He is right, and this PR got here first (2026-05-19 vs 2026-06-15), so I am closing mine in favour of it.

One thing from mine that might be worth folding in here, entirely your call. Both tests on this PR are gated behind `@RunIf(min_cuda_gpus=1)`, so the ordering is only checked on a GPU runner. The subprocess test is the stronger of the two approaches, since it exercises the real `_lazy_init` rather than a stand-in, and I would keep it either way.

The gap is just that on a CPU-only run nothing fails if someone reorders those two lines again. A mock-based ordering check covers that case and needs no GPU:

```python
@mock.patch("lightning.fabric.accelerators.cuda._check_cuda_matmul_precision")
@mock.patch("torch.cuda.set_device")
def test_set_cuda_device_calls_set_device_before_matmul_check(set_device_mock, matmul_check_mock):
    call_order = []
    set_device_mock.side_effect = lambda _: call_order.append("set_device")
    matmul_check_mock.side_effect = lambda _: call_order.append("matmul_check")

    CUDAAccelerator().setup_device(torch.device("cuda", 3))

    assert call_order == ["set_device", "matmul_check"]
```

It only proves call order, not that CUDA actually initialised on the right device, so it is weaker than what you have. It is meant as a cheap guard alongside the subprocess test, not a replacement for it. It ran green on CI in #21767 against the identical source change.

Happy for you to take it, adapt it, or ignore it. Sorry for the duplicate work.

**Review by bhimrazy — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
